### PR TITLE
perf: Quadtree refactor

### PR DESCRIPTION
**Pull request by Dedado3000**
Full rework of quadtree based on Task #865c2x27z
- No more duplicated render
- Remove "Shared GameObject", only object at leaves
- More eficient rendering selected game object

### CODE REVIEW

**Review by gerard-queralt — COMMENTED** [review]
Do we know the cause of #217? Because a GameObject shouldn't appear in the quadtree twice, even before this refactor, or am I wrong? I thought we fixed that.
Also, I don't know if the vector `drawnGameObjects` can be removed, since it was used to prevent a game object from being drawn twice, even if it appeared multiple times in the vector to drawn.

**Comment by Dedado3000** [comment]
> added

In fact #217 was one of the focus of this refactor and is solved with this pr

File: Source/DataStructures/Quadtree.h [hunk 1]
```diff
@@ -60,6 +60,8 @@ class Quadtree
 
 	std::list<GameObject*> GetAllGameObjects(GameObject* gameObject);
 
+	void RecalculateQuadtree();
```
Comment (PaablooCH) [inline Source/DataStructures/Quadtree.h:63]: is this function called somewhere?

**Review by PaablooCH — CHANGES_REQUESTED** [review]
Plus some code refactor and questions, i loaded the first scene and count the gameobject that has (i delete some like the whole blockout father and cameras) and compared them with the size of opaquesToDraw and is like 320~ vs 407, can you check it too?

File: Source/Modules/ModuleRender.h [hunk 2]
```diff
@@ -54,7 +55,7 @@ class ModuleRender : public Module
 
 	unsigned vbo;
 	
-	std::vector<const GameObject*> opaqueGOToDraw;
+	std::set<const GameObject*> opaqueGOToDraw;
```
[the remark this answers is not in the document]
Comment (Dedado3000) [reply Source/Modules/ModuleRender.h:58]: Im looking for efficiency

File: Source/DataStructures/Quadtree.cpp [hunk 3]
```diff
@@ -516,4 +465,31 @@ void Quadtree::LoadOptions(Json& meta) {
...
+
+/*
+Recalculate Quadtree
+Info: the Quadtree in which this function is called get all objects 
+from his structure, delete it childrens and redone the structure
+*/
+void Quadtree::RecalculateQuadtree()
+{
+	std::set<GameObject*> familyGameObjects;
+	this->GetFamilyObjects(familyGameObjects);
```
[the remark this answers is not in the document]
Comment (Dedado3000) [reply Source/DataStructures/Quadtree.cpp]: well, this was deleted bc was unused.

File: Source/DataStructures/Quadtree.h (hunk 1 above)
Comment (Dedado3000) [reply Source/DataStructures/Quadtree.h]: I think that was something that was dropped after some conversation at physics meetings

File: Source/Modules/ModuleRender.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dedado3000) [reply Source/Modules/ModuleRender.h:58]: Okey, as we are deleting every frame and rebuilding is more eficiency using a normal Set, unordered_set is better in a case we will mantain the "vector" across the ticks and we will not rebuild it so frequently

File: Source/DataStructures/Quadtree.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaablooCH) [reply Source/DataStructures/Quadtree.cpp]: > we don't have any standard about this, should we?

I think that we already have it, but i could be wrong, plus we have a task https://app.clickup.com/t/865bph92u where we deleted unnecessary `this->`

File: Source/Modules/ModuleRender.cpp [hunk 4]
```diff
@@ -461,6 +461,40 @@ void ModuleRender::AddToRenderList(GameObject* gameObject)
...
+			continue;
+		}
+		if (gameObject->IsActive())
+		{
+			if (!CheckIfTransparent(gameObject))
+				opaqueGOToDraw.insert(gameObject);
+			else
+			{
+				const ComponentTransform* transform =
+					static_cast<ComponentTransform*>(gameObject->GetComponent(ComponentType::TRANSFORM));
```
Comment (OriolSors) [inline Source/Modules/ModuleRender.cpp:485]: Why are you retrieving the transform again?

File: Source/Modules/ModuleRender.cpp [hunk 5]
```diff
@@ -461,6 +461,40 @@ void ModuleRender::AddToRenderList(GameObject* gameObject)
...
+				float dist = Length(cameraPos - transform->GetGlobalPosition());
+				while (transparentGOToDraw[dist] != nullptr)
+				{
+					float addDistance = 0.0001f;
```
Comment (OriolSors) [inline Source/Modules/ModuleRender.cpp:489]: Maybe we can start putting epsilon offsets like this in a Local configuration or something? Abro debate

File: Source/Modules/ModuleRender.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (PaablooCH) [reply Source/Modules/ModuleRender.h:63]: are u sure about that? for what i found it's the same as map/unordered_map and if we are rebuilding the structure every frame an unordered_set could fit better.

**Review by PaablooCH — DISMISSED** [review]

File: Source/Modules/ModuleRender.cpp (hunk 5 above)
Comment (Dedado3000) [reply Source/Modules/ModuleRender.cpp:489]: this is someone from graphics, you can speak with them but seems ok

File: Source/Modules/ModuleRender.cpp (hunk 4 above)
Comment (Dedado3000) [reply Source/Modules/ModuleRender.cpp:485]: Mental lag, solved

**Review by PaablooCH — APPROVED** [review]

**Review by OriolSors — APPROVED** [review]

**Pull request merged**